sip/sip-004-materialized-view.md

```diff
@@ -0,0 +1,823 @@
+which row in the fork table the block the lives in, its offset in the row, as well as
+the row and offset for its parent (these four values constitute a "fork
+pointer").  This gives the Stacks peer an efficient way
+to identify an ancestor block that is `i` blocks in the past:
+
+1. Find the fork pointer for the current block
+2. Look at the pointer's parent row.  If the parent is within `i` blocks back,
+   then return the parent block header hash.
+3. Otherwise, subtract the length of the fork row from `i`, get the first
+   block in this row, load its fork-pointer, and repeat.
+
+The fork table provides a way to encode a child back-pointer in an ART:  a
+back-pointer is the pair `(back-count, node-pointer)`, where `back-count` is the number
+of blocks back from this ART's block to look, and `node-pointer` is the (disk) pointer
+to the node's data in that block's ART (i.e. an offset in the file that encodes the ART where
+the node's data can be found).
+
+To see an example fork table, consider the following blockchain state:
+
+```
+      d-e-f-g
+     /
+a-b-c
+   \ \
+    \ h-i-j
+     \
+      k-l-m
+```
+
+This blockchain has three distinct forks:  `a-b-c-d-e-f-g`, `a-b-c-h-i-j`, and
+`a-b-c-k-l-m`.  Encoded as a fork table, the fork rows would be:
+
+```
+fork ID | block list
+--------|-----------------------
+0       | [a, b, c, d, e, f, g]
+1       | [h, i, j]
+2       | [k, l, m]
+```
+
+The ancestor table would be:
+
+```
+block | fork ID | index | parent  | parent
+      |         |       | fork ID | index
+------|---------|-------|---------|--------
+a     | 0       | 0     | 0       | 0
+b     | 0       | 1     | 0       | 0
+c     | 0       | 2     | 0       | 1
+d     | 0       | 3     | 0       | 2
+e     | 0       | 4     | 0       | 3
+f     | 0       | 5     | 0       | 4
+g     | 0       | 6     | 0       | 5
+h     | 1       | 0     | 0       | 2
+i     | 1       | 1     | 1       | 0
+j     | 1       | 2     | 1       | 1
+k     | 2       | 0     | 0       | 1
+l     | 2       | 1     | 2       | 0
+m     | 2       | 2     | 2       | 1
+```
+
+The chain tips are straightforward to calculate:  for each fork ID whose parent
+fork ID is the same as the fork ID, they are the blocks who have the highest
+index (if there is only one block in a fork row, then it is obviously the chain tip).
+Clearly, these are `g`, `j`, and `m`.
+
+To see how this works, consider finding the block that is four blocks prior to
+`m`.  To do so, the Stacks peer consults the ancestor table and sees that `m` is
+has fork ID 2 whose block list is `[k, l, m]`.  The block list has only three
+items, so the problem becomes instead finding the block that is one block back
+from `k`'s parent.  From the ancestor table, `k`'s parent is from the fork row
+whose fork ID is 0 and whose index is 1.  This would be `b`, and the fork row
+would be `[a, b, c, d, e, f, g]`.  One block back from `b` is `a`.
```

Comment:
I'm not sure about the meaning of `fork row`, but isn't `[a, b, k, l, m]`?

Reply:
It's `[a, b, c, d, e, f, g]` -- row 0 of the fork table.  This is because the parent of `k` is `b`.  `[a, b, k, l, m]` is not a fork row, because it is not an entry in the fork table.

The fork table is constructed by appending blocks to the end of an existing fork row if possible, and only creating a new fork row if this is impossible.  Supposing that the blocks were mined in order `a,b,c,d,e,f,g,h,i,j,k,l,m`, then the fork table would contain 1 row for `a,b,c,d,e,f,g` because each subsequent block up until `g` was able to be appended to an existing fork row.  When block `h` arrives, it would be put into its own fork row, because it is not a child of `g`.  Blocks `i` and `j` would go into `h`'s fork row, because `i` is the child of `h` and `j` is the child of `i`.  When block `k` arrives, it would be added to its own fork row because it cannot be appended to the fork row containing its parent, `b`.